File: server/src/domain/pass_fail_grade_book.py
"""PassFailGradeBook — pass/fail grading scheme."""
from typing import List
from .grade_book import GradeBook, AssignmentScore, GradeResult


class PassFailGradeBook(GradeBook):
    def __init__(self, passing_threshold: float = 60.0):
        self.passing_threshold = passing_threshold

    @property
    def scheme_name(self) -> str:
        return "pass_fail"

    def calculate_grade(self, scores: List[AssignmentScore]) -> GradeResult:
        if not scores:
            return GradeResult(percentage=0.0, letter_grade="F", grade_points=0.0, display_label="Fail")
        pct = self._weighted_percentage(scores)
        passed = pct >= self.passing_threshold
        return GradeResult(
            percentage=round(pct * 10) / 10,
            letter_grade="P" if passed else "F",
            grade_points=0.0,  # Pass/Fail does not contribute to GPA
            display_label="Pass" if passed else "Fail",
        )

File: artifacts/server/src/domain/PassFailGradeBook.ts
import { GradeBook, AssignmentScore, GradeResult } from "./GradeBook";

export class PassFailGradeBook extends GradeBook {
  constructor(private readonly passingThreshold: number = 60) {
    super();
  }

  get schemeName(): string {
    return "pass_fail";
  }

  calculateGrade(scores: AssignmentScore[]): GradeResult {
    if (scores.length === 0) {
      return { percentage: 0, letterGrade: "F", gradePoints: 0.0, displayLabel: "Fail" };
    }
    const pct = this.weightedPercentage(scores);
    const passed = pct >= this.passingThreshold;
    return {
      percentage: Math.round(pct * 10) / 10,
      letterGrade: passed ? "P" : "F",
      gradePoints: passed ? 0.0 : 0.0, // Pass/Fail doesn't contribute to GPA
      displayLabel: passed ? "Pass" : "Fail",
    };
  }
}
